### PR TITLE
Polytech launch

### PR DESCRIPTION
**Pull request by nikita-klsh**
Modified `dump` action, inference script. Added `shift_pick` action

### CODE REVIEW

File: seismicpro/src/utils.py [hunk 1]
```diff
@@ -802,3 +802,23 @@ def check_unique_fieldrecord_across_surveys(surveys_by_fieldrecord, index):
     """
     if len(surveys_by_fieldrecord) != 1:
         raise ValueError('Field {} represents data from more than one survey!'.format(index))
+
+def transform_pickingstyle_polytech(path, max_len=(6, 4)):
```
Comment (roman-kh) [inline seismicpro/src/utils.py]: Very wrong name

File: seismicpro/src/seismic_batch.py [hunk 2]
```diff
@@ -1253,3 +1259,43 @@ def equalize(self, index, src, dst, params, survey_id_col=None):
...
+        if isinstance(self.index, KNNIndex):
+            trace = trace[0]
+        trace = np.squeeze(trace)
+
+        analytic = hilbert(trace)
+        phase = np.unwrap(np.angle(analytic))
+        # finding x such that phase[x] = phase[pick] - shift
+        phase_mod = phase - phase[pick] + shift
+        phase_mod[phase_mod < 0] = 0
+        zero = len(phase_mod) - phase_mod[::-1].argmin() - 1 # in case phase_mod reaches 0 find the index of last one
```
Comment (a-arefina) [inline seismicpro/src/seismic_batch.py]: why `zero`? please provide more expressive name

File: docker_containers/picking_docker/picking_inference.py [hunk 3]
```diff
@@ -61,6 +62,8 @@ def predict(path_raw, path_model, num_zero, save_to, batch_size, trace_len, devi
         The number of first samples in the trace to load to the pipeline.
     device: str or torch.device, default: 'cpu'
         The device used for inference. Can be 'gpu' in case of avaliavle GPU.
+    shift: int, default: 0
+        Alter the picking times for each trace on the given phase shift. Multiplied by `pi`.
```
Comment (a-arefina) [inline docker_containers/picking_docker/picking_inference.py]:
```suggestion
        Modify the picking times for each trace with the given phase shift. Multiplied by `pi`.
```

File: seismicpro/src/seismic_batch.py [hunk 4]
```diff
@@ -482,13 +483,18 @@ def _dump_picking(self, src, path, traces, to_samples, columns=None):
             df = df.sort_values(by=sort_by)
 
         df = df.loc[self.indices]
-        df['timeOffset'] = data.astype(int)
+
+        if isinstance(self.index, KNNIndex):
+            df = df.iloc[::5, :]
+
+        df[PICKS_FILE_HEADERS[-1]] = data.astype(int)
```
[the remark this answers is not in the document]
Comment (dpodvyaznikov) [reply seismicpro/src/seismic_batch.py]: I'd suggest
```suggestion
        df[PICKS_FILE_HEADERS[2] = data.astype(int)
```

File: seismicpro/src/seismic_batch.py [hunk 5]
```diff
@@ -448,7 +449,7 @@ def _dump_single_segy(self, src, path):
         return self
 
     @action
-    def _dump_picking(self, src, path, traces, to_samples, columns=None):
+    def _dump_picking(self, src, path, traces='raw', to_samples=True, columns=None):
```
Comment (dpodvyaznikov) [inline seismicpro/src/seismic_batch.py]: Remove default for `traces` - this is too implicit

File: seismicpro/src/utils.py [hunk 6]
```diff
@@ -802,3 +806,43 @@ def check_unique_fieldrecord_across_surveys(surveys_by_fieldrecord, index):
...
+    Parameters
+    ----------
+    path : str
+        Path to the file with picking.
+    path_save : str, optional
+        Path where the result would be stored. By default the file would be overwritten.
+    n_spaces : int, default is 8
+        The number of spaces separating columns.
+    max_len : tuple, default is (6, 4)
+        The maximum number of digits each column, except the last one, contains.
```
Comment (dpodvyaznikov) [inline seismicpro/src/utils.py]:
```suggestion
        Width of each column except last one.
```

File: seismicpro/src/utils.py [hunk 7]
```diff
@@ -802,3 +806,43 @@ def check_unique_fieldrecord_across_surveys(surveys_by_fieldrecord, index):
     """
     if len(surveys_by_fieldrecord) != 1:
         raise ValueError('Field {} represents data from more than one survey!'.format(index))
+
+
+def transform_to_fixed_width_columns(path, path_save=None, n_spaces=8, max_len=(6, 4)):
+    """ Transforms the format of the csv file with dumped picking so all the columns are separated
+    by `n_spaces` spaces exactly. To make such transform possible you must provide the maximum number
+    of digits each column, except the last one, contains. In case, for example, traces are identified
+    by the 'FieldRecord' and 'TraceNumber' headers, and they varies from 1 to 123456 and from 1 to 1234 respectively,
```
Comment (dpodvyaznikov) [inline seismicpro/src/utils.py]:
```suggestion
    by the 'FieldRecord' and 'TraceNumber' headers, and they have values from 1 to 123456 and from 1 to 1234 respectively,
```

File: seismicpro/src/seismic_batch.py [hunk 8]
```diff
@@ -1253,3 +1257,44 @@ def equalize(self, index, src, dst, params, survey_id_col=None):
 
         getattr(self, dst)[pos] = equalized_field
         return self
+
+    @action
+    @inbatch_parallel(init='_init_component', target="threads")
+    def shift_pick_phase(self, index, src, src_traces, dst=None, shift=1.5*np.pi, threshold=0.05):
+        """ Shifts picking time stored in `src` component on the given phase along the traces stored in `src_raw`.
```
Comment (dpodvyaznikov) [inline seismicpro/src/seismic_batch.py]:
```suggestion
        """ Shifts picking time stored in `src` component on the given phase along the traces stored in `src_traces`.
```

File: seismicpro/src/seismic_batch.py [hunk 9]
```diff
@@ -1253,3 +1257,44 @@ def equalize(self, index, src, dst, params, survey_id_col=None):
...
+
+        Parameters
+        ----------
+        src : str
+            The batch components to get the data from.
```
Comment (dpodvyaznikov) [inline seismicpro/src/seismic_batch.py]:
```suggestion
            The batch components to get picking from.
```

File: seismicpro/src/seismic_batch.py [hunk 10]
```diff
@@ -1253,3 +1257,44 @@ def equalize(self, index, src, dst, params, survey_id_col=None):
...
+        dst : str
+            The batch components to put the result in.
+        src_traces: str
+            The batch components where the traces are stored.
+        shift: float
+            The amount of phase to shift, default is 1.5 * np.pi which corresponds to transfering picking times
```
Comment (dpodvyaznikov) [inline seismicpro/src/seismic_batch.py]: I suggest to measure `shift` in radians and multiply by `np.pi` in the function body

File: seismicpro/src/seismic_batch.py [hunk 11]
```diff
@@ -1253,3 +1257,44 @@ def equalize(self, index, src, dst, params, survey_id_col=None):
...
+            from 'max' to 'zero' type.
+        threshold: float
+            Threshold determining how many trace samples with low amplitudes, less then thd, can be skipped.
```
Comment (dpodvyaznikov) [inline seismicpro/src/seismic_batch.py]: What does it determine? How many samples can skipped or which amplitudes can be skipped?
Also, what is  thd?

File: seismicpro/src/seismic_batch.py (hunk 9 above)
Comment (dpodvyaznikov) [reply seismicpro/src/seismic_batch.py]: This action does not support multiple components, so change `components` to `component` for other parameters too

File: seismicpro/src/seismic_batch.py [hunk 12]
```diff
@@ -1253,3 +1257,44 @@ def equalize(self, index, src, dst, params, survey_id_col=None):
...
+        pick = getattr(self, src)[pos]
+        trace = getattr(self, src_traces)[pos]
+        if isinstance(self.index, KNNIndex):
+            trace = trace[0]
+        trace = np.squeeze(trace)
+
+        analytic = hilbert(trace)
+        phase = np.unwrap(np.angle(analytic))
+        # finding x such that phase[x] = phase[pick] - shift
+        phase_mod = phase - phase[pick] + shift
```
Comment (dpodvyaznikov) [inline seismicpro/src/seismic_batch.py]:
```suggestion
        phase_mod = phase - (phase[pick] - shift)
```
Just a bit more self-explanatory

File: seismicpro/src/seismic_batch.py [hunk 13]
```diff
@@ -457,38 +458,40 @@ def _dump_picking(self, src, path, traces, to_samples, columns=None):
...
+            data = self.meta[src_traces]['samples'][data]
 
-        if columns is None:
-            columns = PICKS_FILE_HEADERS
+        if PICKS_FILE_HEADER not in columns:
+            columns = columns + (PICKS_FILE_HEADER, )
 
         df = self.index.get_df(reset=False)
-        sort_by = self.meta[traces]['sorting']
+        sort_by = self.meta[src]['sorting']
```
Comment (dpodvyaznikov) [inline seismicpro/src/seismic_batch.py]: Use `get` here

**Pull request merged**